### PR TITLE
Update controller-runtime to 0.7.2

### LINKED ISSUE #267: Update to controller-runtime 0.7.0
**What would you like to be added**:
There was quite a lot of changes especially with introducing client.Object

**Why is this needed**:

### PR DESCRIPTION
**Pull request by alenkacz**
**What this PR does / why we need it**:
Updates dependency to controller-runtime 0.7.2. Most changes are around newly introduced type `client.Object`

Fixes #267

### CODE REVIEW

**Review by jbarrick-mesosphere — APPROVED** [review]

**Comment by kensipe** [comment]
I thought all the kudo -> kuttl references were already did... this is odd...  glad they are getting fixed

**Comment by kensipe** [comment]
main on my machine has no kudo references in code

```
.goreleaser.yml:48:      email: release@kudo.dev
CONTRIBUTING.md:60:Learn how to engage with the KUTTL community on the [community page](https://kudo.dev/community/).
README.md:32:Learn more on how to engage with the KUDO community on the [community page](https://kudo.dev/community/).
keps/0002-operator-testing.md:243:    // The type meta object, should always be a GVK of kudo.dev/v1beta1/TestStep.
keps/0002-operator-testing.md:280:apiVersion: kudo.dev/v1beta1
keps/0002-operator-testing.md:329:    // The type meta object, should always be a GVK of kudo.dev/v1beta1/TestAssert.
keps/0003-unit-under-test-declaration.md:43:apiVersion: kudo.dev/v1alpha1
keps/0003-unit-under-test-declaration.md:55:apiVersion: kudo.dev/v1alpha1
keps/0004-test-composability.md:57:apiVersion: kudo.dev/v1alpha1
keps/0004-test-composability.md:81:apiVersion: kudo.dev/v1alpha1
keps/0004-test-composability.md:91:apiVersion: kudo.dev/v1alpha1
keps/0006-lifecycle.md:315:  - apiVersion: kudo.dev/v1beta1
keps/0006-lifecycle.md:318:  - apiVersion: kudo.dev/v1beta1 
keps/0008-multi-cluster.yaml:41:apiVersion: kudo.dev/v1alpha1
```

Is there a reason all those changes are in this PR?

**Comment by kensipe** [comment]
@alenkacz  you work is off the v0.8.0 tag not off main (it appears)
It would also be great to separate out those concerns... if there is clean up of that magnitude it drown the meaning of the PR.  I would expect a controller-runtime bump to be challenging on its own in this case.

**Comment by kensipe** [comment]
if you don't mind... I needed an example of a PR with a label... I'm not sure if it still applies

**Comment by alenkacz** [comment]
yeah this is still WIP, I need to resolve the test failures and I don't know how for now.

But good point, I should rebase to master 👍

File: pkg/test/utils/kubernetes.go [hunk 1]
```diff
@@ -667,23 +677,23 @@ func NewClusterRoleBinding(apiVersion, kind, name, namespace string, serviceAcco
...
 
 // WithNamespace naively applies the namespace to the object. Used mainly in tests, otherwise
 // use Namespaced.
-func WithNamespace(obj runtime.Object, namespace string) runtime.Object {
-	obj = obj.DeepCopyObject()
+func WithNamespace(obj *unstructured.Unstructured, namespace string) *unstructured.Unstructured {
+	copy := obj.DeepCopy()
 
-	m, _ := meta.Accessor(obj) //nolint:errcheck // runtime.Object don't have the error issues of interface{}
+	m, _ := meta.Accessor(copy) //nolint:errcheck // runtime.Object don't have the error issues of interface{}
```
Comment (ANeumann82) [inline pkg/test/utils/kubernetes.go:689]: I think the //nolint:errcheck is obsolete now, do we still need this? If yes, we should at least adjust the comment to `unstructured.Unstructured`

File: pkg/test/utils/kubernetes.go [hunk 2]
```diff
@@ -701,14 +711,13 @@ func WithStatus(t *testing.T, obj runtime.Object, status map[string]interface{})
 }
 
 // WithKeyValue sets key in the provided object to value.
-func WithKeyValue(obj runtime.Object, key string, value map[string]interface{}) (runtime.Object, error) {
-	obj = obj.DeepCopyObject()
-
+func WithKeyValue(obj *unstructured.Unstructured, key string, value map[string]interface{}) (*unstructured.Unstructured, error) {
+	obj = obj.DeepCopy()
+	// we need to convert to and from unstructured here so that the types in case_test match when comparing
```
Comment (ANeumann82) [inline pkg/test/utils/kubernetes.go:716]: Why do we need the additional convert `ToUnstructured` here? I agree that we need the `FromUnstructured` in Line 723, but it looks like here the object will always be an `Unstructured`, or not? 

I think we could just replace line 717-720 with 
```
content := obj.UnstructuredContent()
```

File: pkg/test/utils/kubernetes.go [hunk 3]
```diff
@@ -870,7 +878,9 @@ func WaitForDelete(c *RetryClient, objs []runtime.Object) error {
 	// Wait for resources to be deleted.
 	return wait.PollImmediate(100*time.Millisecond, 10*time.Second, func() (done bool, err error) {
 		for _, obj := range objs {
-			err = c.Get(context.TODO(), ObjectKey(obj), obj.DeepCopyObject())
+			actual := &unstructured.Unstructured{}
+			actual.SetGroupVersionKind(obj.GetObjectKind().GroupVersionKind())
+			err = c.Get(context.TODO(), ObjectKey(obj), actual)
```
Comment (ANeumann82) [inline pkg/test/utils/kubernetes.go:883]: I've seen this pattern a couple times in the new code, maybe there's a way to unify it? It's not too bad though, I'd be ok with leaving it as is

**Review by ANeumann82 — APPROVED** [review]
Nice work, I've seen a few things that might be improved, but nothing serious.

**Review by jbarrick-mesosphere — APPROVED** [review]

File: pkg/apis/testharness/v1beta1/test_types.go [hunk 4]
```diff
@@ -107,6 +107,8 @@ type TestStep struct {
 type TestAssert struct {
 	// The type meta object, should always be a GVK of  kuttl.dev/v1beta1/TestAssert or kuttl.dev/v1beta1/TestAssert.
 	metav1.TypeMeta `json:",inline"`
+	// Override the default metadata. Set labels or override the test step name.
+	metav1.ObjectMeta `json:"metadata,omitempty"`
```
Comment (kensipe) [inline pkg/apis/testharness/v1beta1/test_types.go:111]: while we use the k8s object structure... do don't actually use it to create objects in a cluster... I can't see that this is used.. is there value or a necessity to add it?  Not opposed.. more curious

File: pkg/test/case_test.go [hunk 5]
```diff
@@ -3,14 +3,15 @@ package test
 import (
 	"testing"
 
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	harness "github.com/kudobuilder/kuttl/pkg/apis/testharness/v1beta1"
+	testutils "github.com/kudobuilder/kuttl/pkg/test/utils"
+
```
Comment (kensipe) [inline pkg/test/case_test.go:10]: this isn't formatted according to our standard... I'm surprise the linters don't catch it.. but new lines sometimes evade the linters

there should be 3 groups:  core, 3rd party, this project ordered in each group

File: pkg/test/step_test.go [hunk 6]
```diff
@@ -34,11 +34,11 @@ func TestStepClean(t *testing.T) {
 	pod2WithNamespace := testutils.NewPod("hello2", testNamespace)
 	pod2WithDiffNamespace := testutils.NewPod("hello2", "different-namespace")
 
-	cl := fake.NewFakeClientWithScheme(scheme.Scheme, pod, pod2WithNamespace, pod2WithDiffNamespace)
+	cl := fake.NewClientBuilder().WithObjects(pod, pod2WithNamespace, pod2WithDiffNamespace).WithScheme(scheme.Scheme).Build()
```
Comment (kensipe) [inline pkg/test/step_test.go:37]: 👏

File: pkg/http/http.go [hunk 7]
```diff
@@ -5,7 +5,7 @@ import (
 	"fmt"
 	"net/url"
 
-	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
```
Comment (kensipe) [inline pkg/http/http.go:8]: ignore this line... but look at line 40 `client := NewClient()`   Now that we import `client` L40 var name of client collides with import package... life would be just wonderful if it didn't :)

File: pkg/test/utils/kubernetes.go [hunk 8]
```diff
@@ -667,23 +677,23 @@ func NewClusterRoleBinding(apiVersion, kind, name, namespace string, serviceAcco
...
+func NewPod(name, namespace string) *unstructured.Unstructured {
 	return NewResource("v1", "Pod", name, namespace)
 }
 
 // WithNamespace naively applies the namespace to the object. Used mainly in tests, otherwise
 // use Namespaced.
-func WithNamespace(obj runtime.Object, namespace string) runtime.Object {
-	obj = obj.DeepCopyObject()
+func WithNamespace(obj *unstructured.Unstructured, namespace string) *unstructured.Unstructured {
+	copy := obj.DeepCopy()
```
Comment (kensipe) [inline pkg/test/utils/kubernetes.go:687]: `copy` collides with built in function..

File: pkg/test/utils/kubernetes.go (hunk 1 above)
Comment (kensipe) [reply pkg/test/utils/kubernetes.go:689]: still needed...

**Review by kensipe — APPROVED** [review]
/lgtm

some much to love!

**Comment by kensipe** [comment]
looks like changes collided with main... normally I would rebase, but didn't want to do that to your branch... merged to main to resolve conflict.

**Pull request merged**